```{r}
if (basename(getwd()) == "downstreamer_main") {setwd("evaluating_coregulation")} else {setwd("../evaluating_coregulation")}
source("../downstreamer_functions.r")
source("data/source_olivier.r")

library(pROC)
library(data.table)
library(scales)


# Cached 2021-04-28
datasets            <- read.downstreamer.batch(main.downstreamer.output.path, USE.CACHE = T)

gene.pvalues        <- read.genep(list.files("../data/final_paper/gene_pvalues/", full.names=T), trim.colnames = c("_hg19_normalizedGenePvalues.txt", "_hg19.txt"))
ol                  <- intersect(names(datasets), colnames(gene.pvalues))


# Gnomad PLI
gnomad <- read.table(gnomad.file, sep="\t", header=T, stringsAsFactors = F)
gnomad <- gnomad[!duplicated(gnomad$gene),]
rownames(gnomad) <- make.names(gsub("\\.\\d+", "", ensembl[gnomad$gene, ]$Gene.stable.ID.version), unique=T)

# HPO matrix
hpo.matrix <- fread(hpo.file, data.table=F)
rownames(hpo.matrix) <- hpo.matrix[,1]
hpo.matrix <- hpo.matrix[,-1]
```



```{r}
# Top 10 AUC, bonfsig, min 20 genes in term
hpo.terms <- c("HP:0002728", "HP:0011370", "HP:0001890", "HP:0001973", "HP:0005403", "HP:0004798", "HP:0020100", "HP:0002841", "HP:0001287", "HP:0002583") 

hpo.term  <- "HP:0002583" # colitis
trait     <- "inflammatory_bowel_disease_2017_29906448"
coreg     <- datasets[[trait]]$GenePrioritization
genep     <- gene.pvalues[,trait, drop=F]

# Overlap genes
ol       <- intersect(intersect(rownames(coreg), rownames(genep)), rownames(hpo.matrix))

results  <- list()

for (hpo.term in hpo.terms) {
  # Calculate AUC
  auc.genep <- roc(as.factor(hpo.matrix[ol, hpo.term]), genep[ol, 1])
  auc.coreg <- roc(as.factor(hpo.matrix[ol, hpo.term]), coreg[ol, ]$Enrichment.Z.score)
  
  results[[hpo.term]] <- list(auc.coreg, auc.genep)
}

#roc.test(auc.genep, auc.coreg)

col1 <- alpha("#376B65", 1)
col2 <- alpha("#D4D952", 1)

plot.roc(results[[1]][[2]],
         col=col1,
         main=paste0(meta.data[trait, "Name"]),
         mgp=c(2.6, 0.7, 0),
         lwd=3,
         las=1)
lines.roc(results[[1]][[1]], col=col2 , lwd = 3)
legend("bottomright",
       legend=c(paste0("Key-genes"),
                paste0("GWAS gene p-values")), 
       col=c(col2, col1), lwd=3, bty="n")

for (hpo.term in hpo.terms[2:length(hpo.terms)]) {
  
  lines.roc(results[[hpo.term]][[2]], col = col1, lwd = 3)
  lines.roc(results[[hpo.term]][[1]], col = col2, lwd = 3)
}

```













# DEPRECATED

# Height 2018
```{r}
dir      <- "~/Desktop/depict2/output/height_paper_v2/"
trait    <- "height_2018_30124842_hg19"
hpo.term <- "HP:0002691" # Platybasia 
hpo.term <- "HP:0001519" # Disproportionate tall stature
hpo.term <- "HP:0004322" # Short stature
base     <- paste0(dir, "/", version, "/", trait)

# Overlap genes
ol       <- intersect(intersect(rownames(coreg), rownames(genep)), rownames(hpo.matrix))

# Calculate AUC
auc.genep <- roc(as.factor(hpo.matrix[ol, hpo.term]), -log10(genep[ol, 1]))
auc.coreg <- roc(as.factor(hpo.matrix[ol, hpo.term]), coreg[ol, ]$Enrichment.Z.score)
```


# IBD 2017
```{r}
dir      <- "~/Desktop/depict2/output/height_paper_v2/"
version  <- "v56"
trait    <- "inflammatory_bowel_disease_2017_29906448_hg19"
hpo.term <- "HP:0005263" # Gastritis
base     <- paste0(dir, "/", version, "/", trait)

# Gene pvalues
genep    <- read.enrichments(paste0(base, "_genePvalues.txt"))
genep    <- genep[!is.na(genep),, drop=F]

# Coregulation Z
coreg    <- read.depict2(paste0(base, "_enrichtments_exHla_1.xlsx"))$Coregulation

# Overlap genes
ol       <- intersect(intersect(rownames(coreg), rownames(genep)), rownames(hpo.matrix))

# Calculate AUC
auc.genep <- roc(as.factor(hpo.matrix[ol, hpo.term]), -log10(genep[ol, 1]))
auc.coreg <- roc(as.factor(hpo.matrix[ol, hpo.term]), coreg[ol, ]$Enrichment.Z.score)
```

# Urate 2021
```{r}
hpo.matrix <- fread("../data/raw_pathway_databases/MGI_2020_10_20_PhenoGenoMP_matrix.txt.gz", data.table=F)
rownames(hpo.matrix) <- hpo.matrix[,1]
hpo.matrix <- hpo.matrix[-1,-1]

results <- read.depict2("../data/pritchard_traits/pritchard_traits_2021_33587031_hg19_enrichtments_Urate.xlsx", potential_traits = c("GenePrioritization200"))$GenePrioritization200
results$GWAS.gene.P.value  <- (1- pnorm(results$GWAS.gene.P.value))*2

hpo.term <- "MP:0011471"
#hpo.term <- "MP:0009643"

ol       <- intersect(rownames(results), rownames(hpo.matrix))

results    <- results[ol,]
hpo.matrix <- hpo.matrix[ol,]

# Gene pvalues
genep    <- results$GWAS.gene.P.value

# Coregulation Z
coreg    <- results$Enrichment.Z.score

# Calculate AUC
auc.genep <- roc(as.factor(hpo.matrix[, hpo.term]), -log10(genep))
auc.coreg <- roc(as.factor(hpo.matrix[, hpo.term]), coreg)

roc.test(auc.genep, auc.coreg)

pdf(width=6.5, height=6.5, file="output/plots/urate_2021_mp_creatinine_in_urine.pdf")
plot.roc(auc.genep,
         col = "goldenrod2",
         main = paste0("Urate GWAS ~ ", " MP: creatinine in urine"),
         mgp=c(2.6, 0.7, 0),
         lwd = 3,
         las=1,
         axes=F)
axis(side=1, at=c(1, 0.5, 0), pos=-0.08)
axis(side=2, at=c(1, 0.5, 0), pos=1.05)
lines.roc(auc.coreg, col = "dodgerblue3", lwd = 3)
legend(x=0.8, y=0.12,
       legend=c(paste0("Urate gene prioritization (AUC: ", round(auc.coreg$auc,2),")"),
                paste0("Urate GWAS gene p-values (AUC: ", round(auc.genep$auc,2),")")), 
       col=c("dodgerblue3", "goldenrod2"), lwd=3, bty="n")
dev.off()

```


# Plot
```{r}
roc.test(auc.genep, auc.coreg)

pdf(width=5, height=5, file=paste0("output/plots/hpo_auc_plot_", trait, "_", hpo.term, ".pdf"))
plot.roc(auc.genep, col = "goldenrod2", main = paste0(trait, " ~ ", hpo.term), mgp=c(2.6, 0.7, 0), lwd = 3, las=1)
lines.roc(auc.coreg, col = "dodgerblue3", lwd = 3)
legend("bottomright",
       legend=c(paste0("IBD Core-gene prioritization (AUC: ", round(auc.coreg$auc,2),")"),
                paste0("IBD GWAS gene p-values (AUC: ", round(auc.genep$auc,2),")")), 
       col=c("dodgerblue3", "goldenrod2"), lwd=3, bty="n")
dev.off()

```
